src/runtime_src/core/pcie/tools/cloud-daemon/container/container.cpp

```diff
@@ -0,0 +1,235 @@
+        return -EINVAL;
+    xclmgmt_ioc_bitstream_axlf obj = {reinterpret_cast<axlf *>(real_xclbin.get()->data())};    
+#endif
+    return mgmtDev->ioctl(XCLMGMT_IOCICAPDOWNLOAD_AXLF, &obj);
+}
+
+bool Container::isGood()
+{
+    return mgmtDev != nullptr;
+}
+
+Container::~Container()
+{
+}
+
+Container::Container(size_t index)
+{
+    mgmtDev = pcidev::get_dev(index, false);
+}
+
+//private methods, vendor dependant
+/*
+ * This file also gives examples how users add customized xclbin protection
+ * leveraging the mpd/msd framework.
+ * 
+ * Entry in the xclbin repository.
+ * This is only for the sample code usage. Cloud vendor has freedom to define
+ * its own in terms of their own implementations
+ */ 
+/* The fake xclbin file transmitted through mailbox is achieved by
+ * #xclbinutil --input path_to_xclbin --remove-section BITSTREAM --output path_to_fake_xclbin
+ * --skip-uuid-insertion
+ * this new fake xclbin has same uuid to the real xclbin
+ *
+ * md5 of the fake xclbin can be achieved by
+ * #md5sum path_to_fake_xclbin
+ *
+ * This md5 is the primary key of the repo database to get the real xclbin
+ */
+struct xclbin_repo {
+    const char *md5; //md5 of the xclbin metadata. should be the primary key of DB of the repo
+    const char *path; //path to xclbin file
+};
+static struct xclbin_repo repo[2] = {
+    {
+        .md5 = "d9662fc2a45422d5f7c80f57dae4c8db",
+        .path = "/opt/xilinx/dsa/xilinx_u280_xdma_201910_1/test/verify.xclbin",
+    },
+    {
+        .md5 = "97aefd0cd3dd9a96cc5d24c9afcd3818",
+        .path = "/opt/xilinx/dsa/xilinx_u280_xdma_201910_1/test/bandwidth.xclbin",
+    },
+}; // there are only 2 xclbins in the sample code
+
+int Container::retrieve_xclbin(const xclBin *&orig,
+       std::shared_ptr<std::vector<char>> &real_xclbin)
+{
+    //go get the real_xclbin yourself
+    char md5[33];
+    calculate_md5(md5, reinterpret_cast<char *>(const_cast<xclBin *>(orig)),
+        orig->m_header.m_length);
+    for (unsigned i= 0; i < sizeof(repo)/sizeof(struct xclbin_repo); i++) {
+        if (strcmp(md5, repo[i].md5) == 0) {
+            return read_file(repo[i].path, real_xclbin);
+        }
+    }
+    return 1;
+}
+
+void Container::calculate_md5(char *md5, char *buf, size_t len)
+{
+    unsigned char s[16];
+    MD5_CTX context;
+    MD5_Init(&context);
+    MD5_Update(&context, buf, len);
+    MD5_Final(s, &context);
+
+    for (int i = 0; i < 16; i++)
+        snprintf(&(md5[i*2]), 3,"%02x", s[i]);
+    md5[33] = 0;
+}
+
+int Container::read_file(const char *filename, std::shared_ptr<std::vector<char>> &sp)
+{
+    std::ifstream t;
+    t.open(filename, std::ios::binary | std::ios::in);
+    if (!t)
+        return 1;
+    t.seekg(0, std::ios::end);
+    int len = t.tellg();
+    t.seekg(0, std::ios::beg);    
+    sp = std::make_shared<std::vector<char>>(len, 0);
+    char *buf = sp.get()->data();
+    t.read(buf, len);
+    t.close();
+    return 0;
+}
```

Comment:
This code should quite smaller. Classical StackOverflow question.
I am puzzled by this shared pointer of vector...
Just use a `std::vector` you construct here and just return it. The move semantics saves you since C++11.

Reply:
In https://github.com/Xilinx/SDAccel_Examples/blob/master/libs/xcl2/xcl2.cpp there is a slightly simpler example.

Reply:
will change to follow the example.